arrays/index.js

```diff
@@ -0,0 +1,170 @@
+  { name: "bread", quantity: 1, purchased: false, price: 25 },
+  { name: "cake", quantity: 1, purchased: false, price: 155 },
+];
+const listWithTotal = myList.map((listItem) => ({
+  ...listItem,
+  total: listItem.quantity * listItem.price,
+}));
+console.log(listWithTotal);
+
+function sortMyList(arr) {
+  return [...arr].sort((a, b) => a.purchased - b.purchased);
+}
+
+function markAsPurchased(arr, product) {
+  const neededIndex = arr.findIndex((arrItem) => arrItem.name == product);
+  if (neededIndex === -1) {
+    return "No such product";
+  } else {
+    arr[neededIndex].purchased = true;
+    return arr;
+  }
+}
+console.log(markAsPurchased(myList, "bread"));
+console.log(sortMyList(myList));
+
+//Видалення продукту зі списку (видалення повинно проводитися шляхом
+//створення нового масиву, в якому продукт, що ми шукаємо, буде відсутнім)
+
+function removeProduct(arr, prodToRemove) {
+  const reserveArr = [...arr];
+  let arrWithoutProduct = [];
+  if (reserveArr.findIndex((item) => item.name === prodToRemove) === -1) {
+    return "The product is not found in the list";
+  } else {
+    arrWithoutProduct = reserveArr
+      .filter((item) => item.name !== prodToRemove)
+      .reduce((total, amount) => {
+        total.push(amount);
+        return total;
+      }, []);
+  }
+
+  return arrWithoutProduct;
+}
+console.log(removeProduct(myList, "cola"));
+
+//Додавання покупки в список. Враховуй, що при додаванні покупки з уже існуючим в списку
+//продуктом, необхідно збільшувати кількість в існуючій покупці, а не додавати нову. При цьому
+//також повинна змінитися сума, наприклад, якщо ціна за одиницю 12, а кількості товарів стало 2,
+//то сума буде 24.
+function addProduct(arr, productToAdd) {
+  let reserveArr = [...arr];
+  let reserveArrAdded = [];
+  if (reserveArr.findIndex((item) => item.name === productToAdd.name) === -1) {
+    reserveArr.push(productToAdd);
+    return reserveArr;
+  } else {
+    const idx = reserveArr.findIndex((item) => item.name === productToAdd.name);
+    const newQuantity = reserveArr[idx].quantity + productToAdd.quantity;
+    newPrice = reserveArr[idx].price + productToAdd.price;
+    reserveArrAdded = reserveArr
+      .slice(0, idx)
+      .concat(reserveArr.slice(idx + 1, reserveArr.length))
+      .concat(
+        reserveArr
+          .filter((item) => item.name === productToAdd.name)
+          .map((item) => {
+            if (item.name == productToAdd.name)
+              return { ...item, quantity: newQuantity, price: newPrice };
+          })
+      );
+
+    return reserveArrAdded;
+  }
+}
+console.log(
+  addProduct(myList, {
+    name: "cake",
+    quantity: 1,
+    purchased: false,
+    price: 155,
+  })
+);
+
+//Підрахунок суми всіх продуктів (враховуючи кількість кожного) в списку.
+function totalSum(arr) {
+  const reserveArr = [...arr];
+  return reserveArr.reduce(
+    (total, amount) => total + amount.quantity * amount.price,
```

Comment:
as from methods above i assumed that the price is the price for all products so there is not need to multiply by quantity